Add CI with lint and format tasks + configure lint-staged

thdk opened the pull request:
This pull request introduces several improvements to the project, including setting up CI/CD workflows, enforcing code quality standards, updating configurations, and enhancing code consistency. The most significant changes include the addition of a GitHub Actions workflow for CI, integration of ESLint and Prettier for linting and formatting, and various code refinements for better readability and maintainability.

### CI/CD and Automation Setup:
* Added a GitHub Actions workflow (`.github/workflows/ci.yml`) to automate build, linting, formatting checks, and tests on `push` and `pull_request` events for the `main` branch.
* Configured Husky pre-commit hooks (`.husky/pre-commit`) to run `lint-staged` for staged file checks.

### Code Quality and Tooling:
* Introduced ESLint and Prettier configurations (`eslint.config.js`, `prettier.config.js`) for enforcing consistent code style and formatting. [[1]](diffhunk://#diff-a32a0887ed9d1d707bbb3b845b7df7fd40e673c47e7b60a3ebd896b68d3b8839R1-R99) [[2]](diffhunk://#diff-e42472dd65561fe56344281078077fb924cddca92821eadfbc2318cca3f89c9aR1-R12)
* Updated `package.json` with new scripts for linting and formatting (`lint`, `format`, `format:check`) and added related dependencies like `eslint`, `prettier`, and `husky`. [[1]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L15-R18) [[2]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R53-R73)

### Codebase Enhancements:
* Improved error logging formatting across multiple functions in `src/index.ts` for better readability and debugging. [[1]](diffhunk://#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80L34-R50) [[2]](diffhunk://#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80L52-R83) [[3]](diffhunk://#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80L74-R110) [[4]](diffhunk://#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80L92-R143) [[5]](diffhunk://#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80L114-R170) [[6]](diffhunk://#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80L132-R181) [[7]](diffhunk://#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80L142-R194)
* Applied consistent trailing commas and formatting in various files, including `src/config.ts` and `src/jira.ts`. [[1]](diffhunk://#diff-c3095d5010e65c52737a98a5d618ea24049ebe90c8470752426081d70ed6e012L44-R44) [[2]](diffhunk://#diff-01d7efd6eea28cdd9969b12f2f6d2d0e1729d4ffa2fc8d93db8cc8c3538c01b3L9-R10) [[3]](diffhunk://#diff-01d7efd6eea28cdd9969b12f2f6d2d0e1729d4ffa2fc8d93db8cc8c3538c01b3L19-R20) [[4]](diffhunk://#diff-01d7efd6eea28cdd9969b12f2f6d2d0e1729d4ffa2fc8d93db8cc8c3538c01b3L33-R34) [[5]](diffhunk://#diff-01d7efd6eea28cdd9969b12f2f6d2d0e1729d4ffa2fc8d93db8cc8c3538c01b3L42-R43) [[6]](diffhunk://#diff-01d7efd6eea28cdd9969b12f2f6d2d0e1729d4ffa2fc8d93db8cc8c3538c01b3L61-R64)

### Documentation Updates:
* Updated `README.md` to improve formatting and consistency in code examples and instructions. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R39) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L49-R50) [[3]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L110-R108) [[4]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R125)

### Environment and Configuration:
* Added `.vscode/extensions.json` to recommend useful extensions like Prettier and ESLint for developers.
* Specified Node.js version `24.0.0` in `.tool-versions` for environment consistency.

Fixes ivelin-web/tempo-mcp-server#11

thdk commented:
No lint issues have been fixed with this PR to keep focus on configuration first.
Next PRs can incrementally fix all currently ignored issues.

ivelin-web commented:
Hi @thdk,

After testing your PR locally, I found several issues that need to be addressed before we can merge:

Critical Issues Found:

1. lint-staged Configuration is Broken
"lint-staged": {
  "*.js": "eslint --cache --fix",      // Missing .ts files
  "*.--write": "prettier --write"      // Syntax error - what is "*.--write"?
}
I tested this and it doesn't work - added a console.log and enabled no-console rule. npm run lint correctly shows errors, but when
I stage the file and run git commit, lint-staged passes without catching the lint errors.

2. Husky Deprecation Warning
During installation: husky - install command is DEPRECATED
Your prepare script uses the old Husky API.

3. Node.js Version Mismatch
- .tool-versions: Node.js 24.0.0
- package.json: "node": ">=18.0.0"
- GitHub Actions: Node.js 24

Why did you choose 24.0.0 specifically? Is there any reason? This is causing compatibility warnings during npm install.

4. Prettier Command Syntax Question
"format": "prettier . --write",
"format:check": "prettier . --check"
Shouldn't the dot be at the end? Like prettier --write . and prettier --check .? Or are both syntaxes valid?

5. Scope Issues
Your commands are processing the entire project instead of just source code:
- prettier . formats config files like .claude/settings.local.json
- ESLint checks non-TypeScript files
- Should these only target src/ directory? At least the ESLint checks.

6. ESLint Configuration Questions
- The boolean flags in eslint.config.js seem unusual:
const weAreHighlyProficientInTypeScript = true;
const weLikeCodeStylingConsistencyEvenIfSomeRulesAreOpiniated = true;
const weDontMindSlowerRulesAsTheyAreUsuallyTheBestBecauseTheyUseTypeInformation = true;
- Is this standard practice or your custom approach?

7. lint-staged Process Flow
Shouldn't the flow be:
1. Run ESLint with --fix
2. Run Prettier
3. Run ESLint again (without --fix) to catch remaining issues and fail if any exist?

Currently, after auto-fixing, there's no final validation step.

Could you clarify these points and fix the broken configurations? The concept is great, but the implementation needs refinement.

Thanks!

thdk commented:
> Hi @thdk,
> 
> After testing your PR locally, I found several issues that need to be addressed before we can merge:
> 
> Critical Issues Found:
> 
> 1. lint-staged Configuration is Broken
>    "lint-staged": {
>    "_.js": "eslint --cache --fix",      // Missing .ts files
>    "_.--write": "prettier --write"      // Syntax error - what is "*.--write"?
>    }
>    I tested this and it doesn't work - added a console.log and enabled no-console rule. npm run lint correctly shows errors, but when
>    I stage the file and run git commit, lint-staged passes without catching the lint errors.

Fixed that. I think I must have tested it on a js file.


> 2. Husky Deprecation Warning
>    During installation: husky - install command is DEPRECATED
>    Your prepare script uses the old Husky API.

Updated. 

> 3. Node.js Version Mismatch
> 
> * .tool-versions: Node.js 24.0.0
> * package.json: "node": ">=18.0.0"
> * GitHub Actions: Node.js 24
> 
> Why did you choose 24.0.0 specifically? Is there any reason? This is causing compatibility warnings during npm install.
> 

The version in `.tool-versions` specifies only the version to use for development. If you are using `asdf` it will automatically use node 24 for building the library.

The version range in package.json specified >= 18.0.0. So 24 is a valid node version according to that.
I don't get warnings when running `npm install` on node 24.

Can you share the warning you get?

> 4. Prettier Command Syntax Question
>    "format": "prettier . --write",
>    "format:check": "prettier . --check"
>    Shouldn't the dot be at the end? Like prettier --write . and prettier --check .? Or are both syntaxes valid?

Prettier docs also use both syntaxes in there docs here: https://prettier.io/docs/cli 

> 5. Scope Issues
>    Your commands are processing the entire project instead of just source code:
> 
> * prettier . formats config files like .claude/settings.local.json

Yes, if that file is committed I would expect it to use the same formatting as every other json file in the repository.

> * ESLint checks non-TypeScript files
> * Should these only target src/ directory? At least the ESLint checks.

I usually lint every file except for those in the `ignores` section of the eslint configuration. Lot's of configuration files are often in the root instead of in `src`.  You do want to lint those files.
> 
> 6. ESLint Configuration Questions
> 
> * The boolean flags in eslint.config.js seem unusual:
>   const weAreHighlyProficientInTypeScript = true;
>   const weLikeCodeStylingConsistencyEvenIfSomeRulesAreOpiniated = true;
>   const weDontMindSlowerRulesAsTheyAreUsuallyTheBestBecauseTheyUseTypeInformation = true;
> * Is this standard practice or your custom approach?

This is an approach that allows me to just copy / paste the `eslint.config.js` file from any project to any project and I just have to toggle the flags depending on the requirements. It also allows the rules to be easily switched into a more relaxed rule set without requiring any knowledge of eslint setup. The flag names serve as documentation both at declaration and at usage to inform anyone immediately that there are possibilities to make it more relaxed/strict.

I've also shared this configuration here:  https://medium.com/@t.dekiere/my-favorite-eslint-setup-9d63ab4de470

But of course feel free to remove the flags and only keep the relevant config parts. Some rules though can be really strict, so just be aware that you now have an easy way out if things get too hard to fix.

> 
> 7. lint-staged Process Flow
>    Shouldn't the flow be:
> 8. Run ESLint with --fix
> 9. Run Prettier
> 10. Run ESLint again (without --fix) to catch remaining issues and fail if any exist?
> 
> Currently, after auto-fixing, there's no final validation step.

I don't see why you would need a final validation step? Unless if you expect one eslint rule fix to introduce another eslint rule error? If so, you still have your CI which will check all files for lint / format issues.

> Thanks!

Thanks for reviewing! 👍🏼

ivelin-web commented:
Hi @thdk, great work on the setup! All configuration issues have been resolved and the tooling works perfectly. Thanks for being responsive to feedback and implementing modern best practices. 🚀
